src/geff/rustworkx/io.py

```diff
@@ -0,0 +1,264 @@
+    edge_props = list({k for _, data in edge_data for k in data})
+
+    write_dicts(
+        geff_store=store,
+        node_data=node_data,
+        edge_data=edge_data,
+        node_prop_names=node_props,
+        edge_prop_names=edge_props,
+        axis_names=axis_names,
+        zarr_format=zarr_format,
+    )
+
+    # write metadata
+    roi_min: tuple[float, ...] | None
+    roi_max: tuple[float, ...] | None
+    if axis_names is not None and graph.num_nodes() > 0:
+        roi_min, roi_max = get_roi_rx(graph, axis_names)
+    else:
+        roi_min, roi_max = None, None
+
+    axes = axes_from_lists(
+        axis_names,
+        axis_units=axis_units,
+        axis_types=axis_types,
+        roi_min=roi_min,
+        roi_max=roi_max,
+    )
+
+    metadata = create_or_update_metadata(
+        metadata,
+        isinstance(graph, rx.PyDiGraph),
+        axes,
+    )
+    metadata.write(group)
+
+
+def _ingest_dict_rx(graph_dict: GraphDict) -> tuple[rx.PyGraph, dict[int, int]]:
+    """Convert a GraphDict to a rustworkx graph."""
+    try:
+        import rustworkx as rx
+    except ImportError as e:
+        raise ImportError(
+            "rustworkx is not installed. Please install it with `pip install geff[rx]`."
+        ) from e
+
+    metadata = graph_dict["metadata"]
+
+    graph = rx.PyDiGraph() if metadata.directed else rx.PyGraph()
+    graph.attrs = metadata.model_dump()
+
+    # Add nodes with populated properties
+    node_ids = graph_dict["nodes"].tolist()
+    node_props: list[dict[str, Any]] = [{} for _ in node_ids]
+
+    # Populate node properties first
+    indices = np.arange(len(node_ids))
+
+    for name, prop_dict in graph_dict["node_props"].items():
+        values = prop_dict["values"]
+        if "missing" in prop_dict:
+            current_indices = indices[~prop_dict["missing"]]
+            values = values[current_indices]
+        else:
+            current_indices = indices
+
+        values = values.tolist()
+        current_indices = current_indices.tolist()
+
+        for idx, val in zip(current_indices, values, strict=True):
+            node_props[idx][name] = val
+
+    # Add nodes with their properties
+    rx_node_ids = graph.add_nodes_from(node_props)
+
+    # Create mapping from geff node id to rustworkx node index
+    to_rx_id_map = dict(zip(node_ids, rx_node_ids, strict=False))
+
+    # Add edges if they exist
+    if len(graph_dict["edges"]) > 0:
+        # converting to local rx ids
+        edge_ids = np.vectorize(to_rx_id_map.__getitem__)(graph_dict["edges"])
+        # Prepare edge data with properties
+        edges_data: list[dict[str, Any]] = [{} for _ in edge_ids]
+        indices = np.arange(len(edge_ids))
+
+        for name, prop_dict in graph_dict["edge_props"].items():
+            values = prop_dict["values"]
+            if "missing" in prop_dict:
+                current_indices = indices[~prop_dict["missing"]]
```

Comment:
Make an issue about doing this for networkx as well to improve efficiency please :)

Reply:
🎊 👍